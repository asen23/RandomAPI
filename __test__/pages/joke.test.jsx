import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import Joke from "pages/joke/index"

const dummySingle = [
    { joke: "a very interesting jokes" },
    { joke: "a very lame jokes" },
]

const dummyDouble = [
    {
        setup: "a setup for a jokes",
        delivery: "a delivery for a jokes",
    },
    {
        setup: "a setup for a great jokes",
        delivery: "a delivery for a great jokes",
    },
]

describe("Joke", () => {
    beforeEach(() => {
        fetch.resetMocks()
    })

    it("renders single joke unchanged", async () => {
        fetch.mockResponseOnce(JSON.stringify(dummySingle[0]))

        const { asFragment } = render(<Joke />)
        expect(asFragment()).toMatchSnapshot()

        await screen.findByText(dummySingle[0].joke)

        expect(asFragment()).toMatchSnapshot()
    })

    it("renders double joke unchanged", async () => {
        fetch.mockResponseOnce(JSON.stringify(dummyDouble[0]))

        const { asFragment } = render(<Joke />)
        expect(asFragment()).toMatchSnapshot()

        await screen.findByText(dummyDouble[0].setup)
        await screen.findByText(dummyDouble[0].delivery)

        expect(asFragment()).toMatchSnapshot()
    })

    it("gives new single joke", async () => {
        fetch.mockResponses(...dummySingle.map((e) => JSON.stringify(e)))

        render(<Joke />)

        const firstJoke = await screen.findByText(dummySingle[0].joke)
        expect(firstJoke).toBeInTheDocument()

        const button = screen.getByRole("button")
        userEvent.click(button)

        const secondJoke = await screen.findByText(dummySingle[1].joke)
        expect(secondJoke).toBeInTheDocument()
    })

    it("gives new double joke", async () => {
        fetch.mockResponses(...dummyDouble.map((e) => JSON.stringify(e)))

        render(<Joke />)

        const firstSetup = await screen.findByText(dummyDouble[0].setup)
        expect(firstSetup).toBeInTheDocument()
        const firstDelivery = await screen.findByText(dummyDouble[0].delivery)
        expect(firstDelivery).toBeInTheDocument()

        const button = screen.getByRole("button")
        userEvent.click(button)

        const secondSetup = await screen.findByText(dummyDouble[1].setup)
        expect(secondSetup).toBeInTheDocument()
        const secondDelivery = await screen.findByText(dummyDouble[1].delivery)
        expect(secondDelivery).toBeInTheDocument()
    })
})
