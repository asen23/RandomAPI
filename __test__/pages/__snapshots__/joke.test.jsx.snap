// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joke renders double joke unchanged 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center flex-col min-h-screen gap-8 bg-primary"
  >
    <h1
      class="text-4xl font-bold"
    >
      Random Joke
    </h1>
    <p>
      Loading...
    </p>
    <button
      class="border border-gray-400 px-2 py-1 rounded-lg hover:bg-white"
    >
      Get random joke
    </button>
  </div>
</DocumentFragment>
`;

exports[`Joke renders double joke unchanged 2`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center flex-col min-h-screen gap-8 bg-primary"
  >
    <h1
      class="text-4xl font-bold"
    >
      Random Joke
    </h1>
    <p>
      a setup for a jokes
    </p>
    <p>
      a delivery for a jokes
    </p>
    <button
      class="border border-gray-400 px-2 py-1 rounded-lg hover:bg-white"
    >
      Get random joke
    </button>
  </div>
</DocumentFragment>
`;

exports[`Joke renders single joke unchanged 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center flex-col min-h-screen gap-8 bg-primary"
  >
    <h1
      class="text-4xl font-bold"
    >
      Random Joke
    </h1>
    <p>
      Loading...
    </p>
    <button
      class="border border-gray-400 px-2 py-1 rounded-lg hover:bg-white"
    >
      Get random joke
    </button>
  </div>
</DocumentFragment>
`;

exports[`Joke renders single joke unchanged 2`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center flex-col min-h-screen gap-8 bg-primary"
  >
    <h1
      class="text-4xl font-bold"
    >
      Random Joke
    </h1>
    <p>
      a very interesting jokes
    </p>
    <button
      class="border border-gray-400 px-2 py-1 rounded-lg hover:bg-white"
    >
      Get random joke
    </button>
  </div>
</DocumentFragment>
`;
